Async select

infradust opened the pull request.

scravy commented on s3access/select_spec.py:
```diff
@@ -0,0 +1,232 @@
+    SQL = 'SQL'
+
+
+class InputSerialization(Enum):
```
I am a bit undecided on these ones.

With the advent of `typing` I for one became a huge fan of `Literal['Foo', 'Bar']` which I think makes definitions much more succinct and feels more pythonic than those javaesque Enums.

infradust replied:
I don't mind changing these to literals, though i find auto-completion to be easier in this way.

scravy replied:
I do not mind either way.

scravy commented on s3access/s3async/s3select.py:
```diff
@@ -0,0 +1,144 @@
+            print(bucket, key, details)
+    bucket = 'some-bucket'
+    keys = ['key/key1.parquet', 'key/key2.parquet', ...]
+    columns = ['col1', 'col2', 'col5', 'col10']  # some columns all files share
```
How are the types resolved?

infradust replied:
s3select (async) does not provide type resolution as it is providing the lowest level it can for selecting data out of s3  (with a bit of specialization in parquet selection).
In my opinion this allows for better compos-ability with other/different code (hence the pandas specialization file).
one can choose simply to dump the csv blob, read it using csv package or whatever else he wishes to do with bytes.

scravy replied:
fair enough.

scravy commented on s3access/select_spec.py:
```diff
@@ -0,0 +1,232 @@
+    example usage:
+    ---------------------------------------------
+    columns = ['col1', 'col2', 'col5', 'col10']  # some columns all files share
+    filters = [('col1', '=', 'some string value'), ('col5', '>=', 10.2), ('col5', '<', 20.3)]
```
How would one do `IN`?

infradust replied:
That's the easiest one to answer so ...
This is modeled basically (in a much more simplistic way) to the way filtering is done at the pyarrow level ([DNF filters](https://arrow.apache.org/docs/python/generated/pyarrow.parquet.ParquetDataset.html)).
So for IN operator: 
```python
('col1', 'IN', ['a', 'b', 'b'])
```

scravy replied:
✅

scravy reviewed (commented):
I am unsure what to make of this.

The intention of s3access is to have a very simple interface to s3 which allows for quickly selecting dataframes from it, without much ceremony. The selling point of it is:

```
    df: pd.DataFrame = s3.select(
        's3://my-bucket/my-path',
        {
            'dim1': str,
            'dim2': str,
            'metric1': float,
            'metric2': int,
        }, {
            'dim1': 'value',
            'dim2: ['foo', 'bar'],
        })
```

The filters here say "dim1 should equal value" and "dim2 should be one of 'foo' or 'bar' (sql IN)`. More complex filters are possible using the `Condition`.

Moreover, these conditions serve an additional purpose, as s3.select also understands hive partitions in the path (like `s3://bucket/path/dim1=value`) and applies it on them. This makes `s3.select` agnostic of the actual partitioning in s3.

I would imagine an async version of this to provide the very same interface, but `async`. I would also suspect it to support the same features (`IN`, partitions in the path).

This here on the other hand is a completely different library. It is a wrapper over aiobotocore which gives a more type-safe interface modelled in a lot of enums and handles the task of multiplexing several requests at once.

scravy commented:
All checks have failed, this is due to the .github actions not installing the optional dependencies (the workflow can be fixed by making the `pip install` step install those too)

infradust commented:
> I am unsure what to make of this.
> 
> The intention of s3access is to have a very simple interface to s3 which allows for quickly selecting dataframes from it, without much ceremony. The selling point of it is:
> 
> ```
>     df: pd.DataFrame = s3.select(
>         's3://my-bucket/my-path',
>         {
>             'dim1': str,
>             'dim2': str,
>             'metric1': float,
>             'metric2': int,
>         }, {
>             'dim1': 'value',
>             'dim2: ['foo', 'bar'],
>         })
> ```
> 
> The filters here say "dim1 should equal value" and "dim2 should be one of 'foo' or 'bar' (sql IN)". More complex filters are possible using the `Condition`.
> 
> Moreover, these conditions serve an additional purpose, as s3.select also understands hive partitions in the path (like `s3://bucket/path/dim1=value`) and applies it on them. This makes `s3.select` agnostic of the actual partitioning in s3.
> 
> I would imagine an async version of this to provide the very same interface, but `async`. I would also suspect it to support the same features (`IN`, partitions in the path).
> 
> This here on the other hand is a completely different library. It is a wrapper over aiobotocore which gives a more type-safe interface modelled in a lot of enums and handles the task of multiplexing several requests at once.

I admit I went a bit out of the way here to better understand/encompass the s3-select.
I was not trying to mimic/duplicate the current api, but to provide a low-level calls to be compose later on.
I was trying to have a separation between the low-level calls and the actual data exposure to the user (pandas is great but could be made an optional requirement for this package).
I went in a different direction with filtering as i found it too obscure as to the actual comparison being done by s3.select in expressing conditions like: 
```python
"'2021-01-01' <= s.date AND s.date < '2021-02-01'"
```
was not intuitive and might direct one to use an 'IN' operator instead (and I've seen it done).

the selling point here is probably the s3pandas (async) which makes it more pandas like to convert csv to a dataframe:
```python
df: pd.DataFrame = await s3pandas.dataframe_from_parquets(
  client, 
  {'some-bucket': ['keys', ...]}, # this makes it flexible in loading from different buckets and files (as long as all involved columns exist)
  columns=['dim1', 'dim2', 'metric1', 'metric2'],
  dtypes={'dim1': 'category', 'metric1': float, 'metric2': float}, # this does not require to specify all column types and exposes the pandas types which greatly reduce memory usage
  filters=[('dim1', '=', 'value'), ('dim2', 'IN', ['foo', 'bar'])]) # i find this filtering to be more transparent, though I would like an Expression builder to expose all possible filtering options (or use DNF like pyarrow)
```
the main difference here might be the loose definition of Sources.
for me, this feels like it should be another function wrapping the 'dataframe_from_parquets' (something like 'dataframe_from_parquets_glob(client, str, ...)' (or event provide the glob functionality as a function 'glob_sources(str)'. 
this would keep things flexible/composable.

the use of aiobotocore could be avoided (though it is probably uses async all the way down which will give it a better performance overall), and indeed in the first versions I simply had a pool executor running the select directly from botocore.
However the progress_callback would be harder to support if using threads (using a queue or other sync mechanism which is not needed in pure async implementation).

This is only an initial/exploratory contribution just to have the option of an async implementation and maybe better structurize the s3-select specification.
there is much more work that needs to go into filters and pandas additional arguments support, and maybe other parts of the s3-select api itself (json to pandas?).

about the failing build ... not really sure what i should change and where ...

scravy commented:
> about the failing build ... not really sure what i should change and where ...

You would want to edit https://github.com/scravy/s3access/blob/main/.github/workflows/python-app.yml#L35

scravy commented:
@infradust Till version v0.5.1 some changes/improvements were introduced which would be worth incorporating into this pull too.

- The filters now allow for `None` too which is translated into `IS NULL` accordingly
- A framework for readers has been introduced which this pull might benefit from too

I would like to incorporate this into s3access, but I would also not just have the low-level calls but a high-level api with the same functionality as the existing one.

I am wondering how to deal with the optional dependency on aiobotocore. I would like to keep it optional, but smth like try import ... except log aiobotocore not present - no async for you (as I would like to simply have the s3 object expose async methods too, WDYT).

infradust commented:
> @infradust Till version v0.5.1 some changes/improvements were introduced which would be worth incorporating into this pull too.
> 
> * The filters now allow for `None` too which is translated into `IS NULL` accordingly
> * A framework for readers has been introduced which this pull might benefit from too
> 
> I would like to incorporate this into s3access, but I would also not just have the low-level calls but a high-level api with the same functionality as the existing one.
> 
> I am wondering how to deal with the optional dependency on aiobotocore. I would like to keep it optional, but smth like try import ... except log aiobotocore not present - no async for you (as I would like to simply have the s3 object expose async methods too, WDYT).

I've added some changes and incorporated the async call into the S3Access object (with the same signature and functionality).
I've also added the option to load pandas dataframes with categorical and date/datetime columns.
This new functionality changed the reader types signature a bit (from 'Dict[str, Type]' to 'Dict[str, Union[Type, str]]' as pandas does not have a proper type for 'categorical).
Let me know what you think.

scravy reviewed (changes requested):
**ci build**
The pipeline/github actions build is still failing. The build is divided into two parts: (1) install dependencies (2) run tests. Step (1) fails due to the removal of the requirements.txt. The requirements.txt were in here specifically such that the build would install and test exactly those requirements.

**tox**
The addition of tox is superfluous and does not have anything to do with adding async select. AFAICS tox solves a problem that github actions already solves – it has matrix builds and supports different build environments. I would unterstand if you wanted to add tox in a separate pull request such that one can test with different python versions locally. However, the current incarnation of the tox config specifies using python 3.8 (`py38`). Now correct me if I'm wrong, but if I'm not mistaken tox will use python 3.8 in every build, that is the github action testing python 3.7 will now also test 3.8. There is a gh tox actions which makes them play together nice: https://github.com/ymyzk/tox-gh-actions – however, this should be part of a separate pull.

**reader → output serialization**
The `Reader` determines what the output serialization should be. For example the JSON reader does not need to parse csv and emit json, when it can readily accept json from the API. AFAICS this is missing from the pull request. Did you by any chance test the async implementation with the JSON Reader?

infradust commented:
I've removed tox and reinstated requirements.txt
I've made the necessary changes to support different readers.

scravy commented:
Great, thank you for the addition. Merging!

scravy commented:
@infradust I will release this as 0.6.0 but the pipeline for python 3.9 hangs while installing dependencies. I merged anyhow as I believed this to be a coincidence, but it happens again and again. I'll investigate when I have the time and see it fixed.

infradust commented:
> @infradust I will release this as 0.6.0 but the pipeline for python 3.9 hangs while installing dependencies. I merged anyhow as I believed this to be a coincidence, but it happens again and again. I'll investigate when I have the time and see it fixed.

Strange ... I will try and check locally

infradust commented:
> > @infradust I will release this as 0.6.0 but the pipeline for python 3.9 hangs while installing dependencies. I merged anyhow as I believed this to be a coincidence, but it happens again and again. I'll investigate when I have the time and see it fixed.
> 
> Strange ... I will try and check locally

@scravy Looks like the pandas dependencies resolution takes a long time under python 3.9, when I change the requirements to >= 1.2.0 it moves much faster.

scravy commented:
@infradust nice find. I'll up that.
